File: go/mux.go
// Stream multiplexing over one reliable transport: carry many independent logical connections
// ("streams") over a single io.ReadWriteCloser — in practice a RedundantStream, so a machine holds
// ONE bundle of N redundant links and multiplexes all of its connections over it, instead of paying
// N links per connection.
//
// This is the missing half of the redundant primitive: the redundant stream gives you one
// never-interrupted ordered byte pipe; the mux Session gives you as many independent
// io.ReadWriteCloser streams as you want on top of it. A yamux/smux/HTTP-2-style design: SYN opens a
// stream, DATA carries bytes, FIN half-closes, RST aborts, and a per-stream credit window
// (WINDOW_UPDATE) provides flow control so one stream can neither starve the others nor buffer
// without bound. DATA is chunked so a long write cannot monopolise the shared transport.
//
// Session-level ordering caveat, stated plainly: the streams share one underlying ordered byte
// stream, so a stalled transport blocks all of them together (head-of-line coupling, like HTTP/2
// over TCP). That is the deliberate trade for a bounded link count; it is fine when the payload is
// small and the transport is a redundant stream that rarely stalls.
//
// Wire framing over the transport (all integers big-endian), identical in Go and JVM:
//
//	type:u8 | streamID:u32 | length:u32 | payload[length]
//
//	SYN           0x01  length 0            open streamID
//	DATA          0x02  length n            n bytes for streamID
//	FIN           0x03  length 0            no more data from this side of streamID
//	RST           0x04  length 0            abort streamID immediately
//	WINDOW_UPDATE 0x05  length 4            grant streamID this many more bytes of send credit
//
// Stream IDs: the client (session opener) uses odd IDs, the server uses even, so both may open
// without colliding.

package multipath

import (
	"encoding/binary"
	"errors"
	"io"
	"sync"
)

const (
	muxSyn    = 0x01
	muxData   = 0x02
	muxFin    = 0x03
	muxRst    = 0x04
	muxWindow = 0x05

	muxHdrLen    = 1 + 4 + 4
	muxMaxChunk  = 16 * 1024  // max DATA payload per frame — bounds one stream's grab of the transport
	muxDefWindow = 256 * 1024 // per-stream receive window
)

// ErrSessionClosed is returned once the session (or its transport) is gone.
var ErrSessionClosed = errors.New("multipath: mux session closed")

// ErrStreamReset is returned to the peer of a stream that was RST.
var ErrStreamReset = errors.New("multipath: stream reset")

// Session multiplexes many Streams over one transport.
type Session struct {
	transport io.ReadWriteCloser
	client    bool

	wmu sync.Mutex // serialises frame writes to the transport

	mu       sync.Mutex
	streams  map[uint32]*MuxStream
	nextID   uint32
	accept   chan *MuxStream
	closed   bool
	closeErr error
}

// NewClientSession and NewServerSession wrap a transport (e.g. a *RedundantStream). The two ends must
// agree on who is client — the client opens odd stream IDs, the server even.
func NewClientSession(transport io.ReadWriteCloser) *Session { return newSession(transport, true) }
func NewServerSession(transport io.ReadWriteCloser) *Session { return newSession(transport, false) }

func newSession(transport io.ReadWriteCloser, client bool) *Session {
	s := &Session{
		transport: transport,
		client:    client,
		streams:   make(map[uint32]*MuxStream),
		accept:    make(chan *MuxStream, 64),
	}
	if client {
		s.nextID = 1
	} else {
		s.nextID = 2
	}
	go s.readLoop()
	return s
}

// OpenStream opens a new outbound stream.
func (s *Session) OpenStream() (*MuxStream, error) {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil, s.err()
	}
	id := s.nextID
	s.nextID += 2
	st := newMuxStream(s, id)
	s.streams[id] = st
	s.mu.Unlock()
	if err := s.writeFrame(muxSyn, id, nil); err != nil {
		s.removeStream(id)
		return nil, err
	}
	return st, nil
}

// AcceptStream returns the next inbound stream opened by the peer.
func (s *Session) AcceptStream() (*MuxStream, error) {
	st, ok := <-s.accept
	if !ok {
		return nil, s.err()
	}
	return st, nil
}

// Close tears down the session, its transport, and every stream.
func (s *Session) Close() error {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil
	}
	s.closed = true
	if s.closeErr == nil {
		s.closeErr = ErrSessionClosed
	}
	streams := make([]*MuxStream, 0, len(s.streams))
	for _, st := range s.streams {
		streams = append(streams, st)
	}
	close(s.accept)
	s.mu.Unlock()
	for _, st := range streams {
		st.shutdown(ErrSessionClosed)
	}
	return s.transport.Close()
}

func (s *Session) err() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closeErr != nil {
		return s.closeErr
	}
	return ErrSessionClosed
}

func (s *Session) removeStream(id uint32) {
	s.mu.Lock()
	delete(s.streams, id)
	s.mu.Unlock()
}

func (s *Session) writeFrame(typ byte, id uint32, payload []byte) error {
	hdr := make([]byte, muxHdrLen)
	hdr[0] = typ
	binary.BigEndian.PutUint32(hdr[1:], id)
	binary.BigEndian.PutUint32(hdr[5:], uint32(len(payload)))
	s.wmu.Lock()
	defer s.wmu.Unlock()
	if s.isClosed() {
		return s.err()
	}
	if _, err := s.transport.Write(hdr); err != nil {
		return err
	}
	if len(payload) > 0 {
		if _, err := s.transport.Write(payload); err != nil {
			return err
		}
	}
	return nil
}

func (s *Session) isClosed() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.closed
}

// readLoop demultiplexes frames off the transport into per-stream buffers.
func (s *Session) readLoop() {
	var hdr [muxHdrLen]byte
	for {
		if _, err := io.ReadFull(s.transport, hdr[:]); err != nil {
			s.fail(err)
			return
		}
		typ := hdr[0]
		id := binary.BigEndian.Uint32(hdr[1:])
		n := binary.BigEndian.Uint32(hdr[5:])
		var payload []byte
		if n > 0 {
			payload = make([]byte, n)
			if _, err := io.ReadFull(s.transport, payload); err != nil {
				s.fail(err)
				return
			}
		}
		switch typ {
		case muxSyn:
			s.onSyn(id)
		case muxData:
			s.onData(id, payload)
		case muxFin:
			if st := s.getStream(id); st != nil {
				st.remoteFin()
			}
		case muxRst:
			if st := s.getStream(id); st != nil {
				st.shutdown(ErrStreamReset)
				s.removeStream(id)
			}
		case muxWindow:
			if st := s.getStream(id); st != nil && len(payload) == 4 {
				st.grantSend(binary.BigEndian.Uint32(payload))
			}
		}
	}
}

func (s *Session) onSyn(id uint32) {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return
	}
	if _, exists := s.streams[id]; exists {
		s.mu.Unlock()
		return
	}
	st := newMuxStream(s, id)
	s.streams[id] = st
	s.mu.Unlock()
	select {
	case s.accept <- st:
	default:
		// Accept backlog full: refuse the stream rather than block the read loop.
		s.removeStream(id)
		_ = s.writeFrame(muxRst, id, nil)
	}
}

func (s *Session) onData(id uint32, payload []byte) {
	st := s.getStream(id)
	if st == nil {
		return // unknown/closed stream — drop
	}
	st.deliver(payload)
}

func (s *Session) getStream(id uint32) *MuxStream {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.streams[id]
}

func (s *Session) fail(err error) {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return
	}
	s.closed = true
	s.closeErr = err
	streams := make([]*MuxStream, 0, len(s.streams))
	for _, st := range s.streams {
		streams = append(streams, st)
	}
	close(s.accept)
	s.mu.Unlock()
	for _, st := range streams {
		st.shutdown(err)
	}
	_ = s.transport.Close()
}

// MuxStream is one logical connection within a Session. It implements io.ReadWriteCloser.
type MuxStream struct {
	sess *Session
	id   uint32

	mu        sync.Mutex
	readable  *sync.Cond
	writable  *sync.Cond
	inbox     []byte
	sendWin   int   // credit the peer has granted us (starts at the default receive window)
	remoteEOF bool  // peer sent FIN
	localFIN  bool  // we sent FIN
	err       error // set on reset/session failure
	closed    bool
}

func newMuxStream(sess *Session, id uint32) *MuxStream {
	st := &MuxStream{sess: sess, id: id, sendWin: muxDefWindow}
	st.readable = sync.NewCond(&st.mu)
	st.writable = sync.NewCond(&st.mu)
	return st
}

func (st *MuxStream) deliver(p []byte) {
	st.mu.Lock()
	if st.closed || st.err != nil {
		st.mu.Unlock()
		return
	}
	st.inbox = append(st.inbox, p...)
	st.readable.Broadcast()
	st.mu.Unlock()
}

func (st *MuxStream) remoteFin() {
	st.mu.Lock()
	st.remoteEOF = true
	st.readable.Broadcast()
	st.mu.Unlock()
}

func (st *MuxStream) grantSend(n uint32) {
	st.mu.Lock()
	st.sendWin += int(n)
	st.writable.Broadcast()
	st.mu.Unlock()
}

func (st *MuxStream) shutdown(err error) {
	st.mu.Lock()
	if st.err == nil {
		st.err = err
	}
	st.readable.Broadcast()
	st.writable.Broadcast()
	st.mu.Unlock()
}

// Read returns delivered bytes in order, replenishing the peer's send window as data is consumed.
func (st *MuxStream) Read(p []byte) (int, error) {
	st.mu.Lock()
	for len(st.inbox) == 0 && st.err == nil && !st.remoteEOF {
		st.readable.Wait()
	}
	if len(st.inbox) == 0 {
		err := st.err
		if err == nil && st.remoteEOF {
			err = io.EOF
		}
		st.mu.Unlock()
		return 0, err
	}
	n := copy(p, st.inbox)
	st.inbox = st.inbox[n:]
	st.mu.Unlock()
	// Replenish flow-control credit for what we just consumed.
	_ = st.sess.writeFrame(muxWindow, st.id, u32bytes(uint32(n)))
	return n, nil
}

// Write sends bytes, chunked and flow-controlled; blocks while the peer's window is exhausted.
func (st *MuxStream) Write(p []byte) (int, error) {
	total := 0
	for len(p) > 0 {
		st.mu.Lock()
		for st.sendWin == 0 && st.err == nil && !st.closed && !st.localFIN {
			st.writable.Wait()
		}
		if st.err != nil || st.closed || st.localFIN {
			e := st.err
			if e == nil {
				e = io.ErrClosedPipe
			}
			st.mu.Unlock()
			return total, e
		}
		n := len(p)
		if n > st.sendWin {
			n = st.sendWin
		}
		if n > muxMaxChunk {
			n = muxMaxChunk
		}
		st.sendWin -= n
		st.mu.Unlock()

		if err := st.sess.writeFrame(muxData, st.id, p[:n]); err != nil {
			return total, err
		}
		p = p[n:]
		total += n
	}
	return total, nil
}

// Close half-closes the write side (sends FIN). Reads may continue until the peer's FIN/EOF.
func (st *MuxStream) Close() error {
	st.mu.Lock()
	if st.localFIN || st.closed {
		st.mu.Unlock()
		return nil
	}
	st.localFIN = true
	st.writable.Broadcast()
	st.mu.Unlock()
	return st.sess.writeFrame(muxFin, st.id, nil)
}

// Reset aborts the stream immediately in both directions.
func (st *MuxStream) Reset() error {
	st.mu.Lock()
	st.closed = true
	if st.err == nil {
		st.err = ErrStreamReset
	}
	st.readable.Broadcast()
	st.writable.Broadcast()
	st.mu.Unlock()
	st.sess.removeStream(st.id)
	return st.sess.writeFrame(muxRst, st.id, nil)
}

func u32bytes(v uint32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, v)
	return b
}

File: jvm/src/main/kotlin/app/microteams/multipath/redundant/Mux.kt
// Stream multiplexing over one reliable transport — the JVM peer of go/mux.go, wire-identical so a
// Go end and a JVM end interoperate. Carries many independent logical connections ("streams") over
// a
// single transport (in practice a RedundantStream), so a machine holds ONE bundle of N redundant
// links and multiplexes all its connections over it rather than paying N links per connection.
//
// yamux/smux/HTTP-2-style: SYN opens a stream, DATA carries bytes, FIN half-closes, RST aborts, and
// a
// per-stream credit window (WINDOW_UPDATE) provides flow control so one stream can neither starve
// the
// others nor buffer without bound. DATA is chunked so a long write cannot monopolise the transport.
// The streams share one ordered byte stream, so a stalled transport blocks them together
// (head-of-line coupling) — the deliberate trade for a bounded link count.
//
// Wire framing over the transport (big-endian): type:u8 | streamID:u32 | length:u32 | payload.
//   SYN 0x01 (len 0) | DATA 0x02 (len n) | FIN 0x03 (len 0) | RST 0x04 (len 0) | WINDOW_UPDATE 0x05
// (len 4)
// Client (session opener) uses odd stream IDs, server even.

package app.microteams.multipath.redundant

import java.io.Closeable
import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.thread
import kotlin.concurrent.withLock

internal const val MUX_SYN = 0x01
internal const val MUX_DATA = 0x02
internal const val MUX_FIN = 0x03
internal const val MUX_RST = 0x04
internal const val MUX_WINDOW = 0x05

internal const val MUX_HDR_LEN = 1 + 4 + 4
internal const val MUX_MAX_CHUNK = 16 * 1024
internal const val MUX_DEF_WINDOW = 256 * 1024

/** A transport the mux runs over: a reliable ordered duplex byte stream. RedundantStream is one. */
interface MuxTransport : Closeable {
    fun read(dst: ByteArray, off: Int, len: Int): Int // -1 on EOF

    fun write(src: ByteArray)
}

/** Adapts a RedundantStream to MuxTransport. */
fun RedundantStream.asMuxTransport(): MuxTransport =
    object : MuxTransport {
        override fun read(dst: ByteArray, off: Int, len: Int): Int =
            this@asMuxTransport.read(dst, off, len)

        override fun write(src: ByteArray) {
            this@asMuxTransport.write(src)
        }

        override fun close() {
            this@asMuxTransport.close()
        }
    }

class MuxSessionClosedException : IOException("multipath: mux session closed")

class MuxStreamResetException : IOException("multipath: stream reset")

/** Multiplexes many [MuxStream]s over one transport. */
class MuxSession private constructor(private val transport: MuxTransport, client: Boolean) :
    Closeable {
    private val wlock = ReentrantLock() // serialises frame writes
    private val lock = ReentrantLock()
    private val streams = HashMap<Long, MuxStream>()
    private var nextId: Long = if (client) 1 else 2
    private val accept = LinkedBlockingQueue<MuxStream>()
    @Volatile private var closed = false
    @Volatile private var closeErr: IOException? = null

    companion object {
        @JvmStatic fun client(transport: MuxTransport): MuxSession = MuxSession(transport, true)

        @JvmStatic fun server(transport: MuxTransport): MuxSession = MuxSession(transport, false)
    }

    private val acceptClosed = MuxStream(this, -1) // sentinel to unblock accept() on close

    init {
        thread(isDaemon = true, name = "mux-read") { readLoop() }
    }

    fun openStream(): MuxStream {
        val id: Long
        val st: MuxStream
        lock.withLock {
            if (closed) throw err()
            id = nextId
            nextId += 2
            st = MuxStream(this, id)
            streams[id] = st
        }
        writeFrame(MUX_SYN, id, null, 0, 0)
        return st
    }

    fun acceptStream(): MuxStream {
        val st = accept.take()
        if (st === acceptClosed) throw err()
        return st
    }

    override fun close() {
        val snapshot: List<MuxStream>
        lock.withLock {
            if (closed) return
            closed = true
            if (closeErr == null) closeErr = MuxSessionClosedException()
            snapshot = streams.values.toList()
        }
        accept.offer(acceptClosed)
        snapshot.forEach { it.shutdown(MuxSessionClosedException()) }
        try {
            transport.close()
        } catch (_: Exception) {}
    }

    private fun err(): IOException = closeErr ?: MuxSessionClosedException()

    internal fun removeStream(id: Long) = lock.withLock { streams.remove(id) }

    internal fun writeFrame(type: Int, id: Long, payload: ByteArray?, off: Int, len: Int) {
        val hdr = ByteArray(MUX_HDR_LEN)
        hdr[0] = type.toByte()
        putU32(hdr, 1, id)
        putU32(hdr, 5, len.toLong())
        wlock.withLock {
            if (closed) throw err()
            transport.write(hdr)
            if (len > 0) {
                transport.write(
                    if (off == 0 && len == payload!!.size) payload
                    else payload!!.copyOfRange(off, off + len)
                )
            }
        }
    }

    private fun readLoop() {
        val hdr = ByteArray(MUX_HDR_LEN)
        try {
            while (true) {
                readFully(hdr, MUX_HDR_LEN)
                val type = hdr[0].toInt() and 0xFF
                val id = u32(hdr, 1)
                val n = u32(hdr, 5).toInt()
                var payload: ByteArray? = null
                if (n > 0) {
                    payload = ByteArray(n)
                    readFully(payload, n)
                }
                when (type) {
                    MUX_SYN -> onSyn(id)
                    MUX_DATA -> getStream(id)?.deliver(payload!!)
                    MUX_FIN -> getStream(id)?.remoteFin()
                    MUX_RST -> {
                        getStream(id)?.shutdown(MuxStreamResetException())
                        removeStream(id)
                    }
                    MUX_WINDOW ->
                        if (payload != null && payload.size == 4)
                            getStream(id)?.grantSend(u32(payload, 0))
                }
            }
        } catch (e: Exception) {
            fail(if (e is IOException) e else IOException(e))
        }
    }

    private fun readFully(b: ByteArray, len: Int) {
        var got = 0
        while (got < len) {
            val r = transport.read(b, got, len - got)
            if (r < 0) throw EOFException()
            got += r
        }
    }

    private fun onSyn(id: Long) {
        lock.withLock {
            if (closed || streams.containsKey(id)) return
            val st = MuxStream(this, id)
            streams[id] = st
            accept.offer(st)
        }
    }

    private fun getStream(id: Long): MuxStream? = lock.withLock { streams[id] }

    private fun fail(e: IOException) {
        val snapshot: List<MuxStream>
        lock.withLock {
            if (closed) return
            closed = true
            closeErr = e
            snapshot = streams.values.toList()
        }
        accept.offer(acceptClosed)
        snapshot.forEach { it.shutdown(e) }
        try {
            transport.close()
        } catch (_: Exception) {}
    }
}

/** One logical connection within a [MuxSession]. */
class MuxStream internal constructor(private val sess: MuxSession, private val id: Long) {
    private val lock = ReentrantLock()
    private val readable = lock.newCondition()
    private val writable = lock.newCondition()
    private var inbox = ByteArray(0)
    private var inboxOff = 0
    private var sendWin = MUX_DEF_WINDOW
    private var remoteEof = false
    private var localFin = false
    private var closed = false
    private var error: IOException? = null

    internal fun deliver(p: ByteArray) =
        lock.withLock {
            if (closed || error != null) return
            // compact + append
            if (inboxOff > 0) {
                inbox = inbox.copyOfRange(inboxOff, inbox.size)
                inboxOff = 0
            }
            inbox += p
            readable.signalAll()
        }

    internal fun remoteFin() {
        lock.withLock {
            remoteEof = true
            readable.signalAll()
        }
        cleanupIfClosed()
    }

    // Frees a stream's slot once both sides have half-closed. A cleanly finished stream must not be
    // aborted with RST to reclaim it: the peer may still be draining buffered DATA, and an RST
    // discards it. So both-FIN is the graceful teardown, and reset() stays for the abnormal case.
    private fun cleanupIfClosed() {
        val done = lock.withLock { localFin && remoteEof }
        if (done) sess.removeStream(id)
    }

    internal fun grantSend(n: Long) =
        lock.withLock {
            sendWin += n.toInt()
            writable.signalAll()
        }

    internal fun shutdown(e: IOException) =
        lock.withLock {
            if (error == null) error = e
            readable.signalAll()
            writable.signalAll()
        }

    fun read(dst: ByteArray, off: Int, len: Int): Int {
        lock.withLock {
            while (inbox.size - inboxOff == 0 && error == null && !remoteEof) readable.await()
            val avail = inbox.size - inboxOff
            if (avail == 0) {
                error?.let { throw it }
                return -1 // remote EOF, drained
            }
            val n = minOf(len, avail)
            System.arraycopy(inbox, inboxOff, dst, off, n)
            inboxOff += n
            // replenish the peer's send credit for what we consumed
            sess.writeFrame(MUX_WINDOW, id, u32Bytes(n.toLong()), 0, 4)
            return n
        }
    }

    fun write(src: ByteArray) {
        var pos = 0
        while (pos < src.size) {
            val n: Int
            lock.withLock {
                while (sendWin == 0 && error == null && !closed && !localFin) writable.await()
                if (error != null || closed || localFin)
                    throw (error ?: IOException("multipath: stream closed"))
                n = minOf(src.size - pos, sendWin, MUX_MAX_CHUNK)
                sendWin -= n
            }
            sess.writeFrame(MUX_DATA, id, src, pos, n)
            pos += n
        }
    }

    /** Half-close the write side (FIN); reads may continue until the peer's EOF. */
    fun closeWrite() {
        lock.withLock {
            if (localFin || closed) return
            localFin = true
            writable.signalAll()
        }
        sess.writeFrame(MUX_FIN, id, null, 0, 0)
        cleanupIfClosed()
    }

    /** Abort the stream in both directions (RST). */
    fun reset() {
        lock.withLock {
            closed = true
            if (error == null) error = MuxStreamResetException()
            readable.signalAll()
            writable.signalAll()
        }
        sess.removeStream(id)
        try {
            sess.writeFrame(MUX_RST, id, null, 0, 0)
        } catch (_: Exception) {}
    }

    fun inputStream(): InputStream =
        object : InputStream() {
            private val one = ByteArray(1)

            override fun read(): Int =
                if (this@MuxStream.read(one, 0, 1) < 0) -1 else one[0].toInt() and 0xFF

            override fun read(b: ByteArray, o: Int, l: Int): Int = this@MuxStream.read(b, o, l)
        }

    fun outputStream(): OutputStream =
        object : OutputStream() {
            override fun write(b: Int) = this@MuxStream.write(byteArrayOf(b.toByte()))

            override fun write(b: ByteArray, o: Int, l: Int) =
                this@MuxStream.write(b.copyOfRange(o, o + l))
        }
}

private fun putU32(b: ByteArray, i: Int, v: Long) {
    b[i] = (v ushr 24).toByte()
    b[i + 1] = (v ushr 16).toByte()
    b[i + 2] = (v ushr 8).toByte()
    b[i + 3] = v.toByte()
}

private fun u32(b: ByteArray, i: Int): Long {
    var v = 0L
    for (k in 0 until 4) v = (v shl 8) or (b[i + k].toLong() and 0xFF)
    return v
}

private fun u32Bytes(v: Long): ByteArray {
    val b = ByteArray(4)
    putU32(b, 0, v)
    return b
}
